MAINT:Fix assert raises in `sklearn/preprocessing/tests/`

sameshl opened the pull request:
replaced `assert_raises` and `assert_raises_regex` with `pytest.raises` context manager.

related to #14216

sameshl committed:
fix `assert_raises` and `assert_raises_regex` in test_discretization.py

sameshl committed:
MANIT fix `assert_raises` and `assert_raises_regex` in test_encoders.py

sameshl committed:
MANIT fix `assert_raises` and `assert_raises_regex` in test_label.py

sameshl committed:
MANIT fix `assert_raises` and `assert_raises_regex` in test_data.py

glemaitre commented:
Same as other PR: you can also change `assert_raise_message`

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -1201,53 +1210,59 @@ def test_quantile_transform_check_error():
-                        " greater than the number of samples used. Got"
-                        " 1000 quantiles and 10 samples.",
-                        QuantileTransformer(subsample=10).fit, X)
+    with pytest.raises(ValueError, match="Invalid value for "
```
a nicer way to be able to read it is to move the error message above `pytest.raises`

```python
err_msg = "Invalid value for 'n_quantiles':0."
with pytest.raises(ValueError, match=err_msg
)
```

You can even make a for loop (you should parametrize but let's keep for another PR)

```python
all_err_msg = ("Invalid value for 'n_quantiles': 0.",
               "Invalid value for 'subsample': 0.",
               ...)
all_params = ({'n_quantiles': 0.}, {'subsample': 0.}, {'subsample': 10})
for params, err_msg in zip(all_params, all_err_msg):
    with pytest.raises(ValueError, match=err_msg):
        QuantileTransformer(**params).fit(X)
```

sameshl replied:
Thanks for the suggestion @glemaitre. Will move the errror msg out of `pytest.raises`. Will open another PR for parameterization purpose.

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -1201,53 +1210,59 @@ def test_quantile_transform_check_error():
-    assert_raises_regex(ValueError, "QuantileTransformer only accepts "
-                        "non-negative sparse matrices.",
-                        transformer.fit, X_neg)
+    with pytest.raises(ValueError, match="QuantileTransformer only accepts "
```
Same comments regarding the error message below. Try to move them out of `raises` and when possible you can loop.

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -643,10 +645,10 @@ def test_one_hot_encoder_invalid_params(X_fit, params, err_msg):
-        ValueError,
-        "`drop` should have length equal to the number",
-        enc.fit, [['abc', 2, 55], ['def', 1, 55], ['def', 3, 59]])
+    with pytest.raises(ValueError,
```
move out the message

glemaitre reviewed (dismissed):
Otherwise, looks good.

sameshl committed:
fix `assert_raise_message` in test_label.py

sameshl committed:
fix `assert_raise_message` in test_base.py

sameshl committed:
fix `assert_raise_message` in test_data.py

sameshl committed:
fix `assert_raise_message` in test_discretization.py

added `match=err_msg.replace('(', '\(').replace(')', '\)'))`
because test case failing due as the regexp parameter of
the match method is matched with the re.search function

sameshl committed:
fix linting errors

sameshl committed:
move out error message in `pytest.raises` in test_data.py

sameshl committed:
move out error message in `test_encoders.py`

sameshl commented:
@glemaitre I have made the required changes. Let me know if anything else needs to be done

rth commented on sklearn/preprocessing/tests/test_discretization.py:
```diff
@@ -39,48 +37,52 @@ def test_valid_n_bins():
-    assert_raise_message(ValueError,
-                         "n_bins must be a scalar or array of shape "
-                         "(n_features,).", est.fit_transform, X)
+    err_msg = "n_bins must be a scalar or array of shape (n_features,)."
```
```suggestion
   err_msg = r"n_bins must be a scalar or array of shape \(n_features,\)."
```
then remove the double `.replace` below.

rth commented on sklearn/preprocessing/tests/test_discretization.py:
```diff
@@ -39,48 +37,52 @@ def test_valid_n_bins():
-                         "(n_features,).", est.fit_transform, X)
+    err_msg = "n_bins must be a scalar or array of shape (n_features,)."
+    with pytest.raises(ValueError,
+                       match=err_msg.replace('(', r'\(').replace(')', r'\)')):
```
Same as above

rth commented on sklearn/preprocessing/tests/test_discretization.py:
```diff
@@ -39,48 +37,52 @@ def test_valid_n_bins():
-                         "(n_features,).", est.fit_transform, X)
+    err_msg = "n_bins must be a scalar or array of shape (n_features,)."
+    with pytest.raises(ValueError,
+                       match=err_msg.replace('(', r'\(').replace(')', r'\)')):
```
Same as above

rth commented on sklearn/preprocessing/tests/test_discretization.py:
```diff
@@ -148,10 +152,12 @@ def test_numeric_stability(i):
+               "('onehot', 'onehot-dense', 'ordinal'). "
+               "Got encode='invalid-encode' instead.")
+    with pytest.raises(ValueError,
+                       match=err_msg.replace('(', r'\(').replace(')', r'\)')):
```
same as above

rth commented on sklearn/preprocessing/tests/test_discretization.py:
```diff
@@ -179,10 +185,12 @@ def test_encode_options():
+               "('uniform', 'quantile', 'kmeans'). "
+               "Got strategy='invalid-strategy' instead.")
+    with pytest.raises(ValueError,
+                       match=err_msg.replace('(', r'\(').replace(')', r'\)')):
```
same as above

rth reviewed (commented):
Few comments, appart from that LGTM. Thanks

sameshl committed:
better way of writing match statement

sameshl committed:
fix linting error

sameshl commented:
Thanks for the suggestions @rth. I have made them

sameshl commented:
@rth Let me know if anything else needs to be done.

rth reviewed (approved):
LGTM, thanks @sameshl !
